### ISSUE TITLE
[iOS] Next video does not play while in background

### ISSUE THREAD

**Issue by rorygarand** [TheWidlarzGroup/react-native-video#1735]
# Bug

Only tested in iOS.

If the app is in the foreground, setting a new source will update the player and a new video begins to play.

If the app is in the background, setting a new source does cause the `onLoadStart` and `onLoad` events to fire, but the next video does not play until the app is brought back into the foreground.

## Environment info
```
react: 16.8.6
react-native: 0.60.0
```

Library version: 5.0.0

## Steps To Reproduce

1. Setup a player with background audio, and a video source of your choice.
2. Programmatically change the source to a different video. (Using `setTimeout` is probably the best bet due to the nature of this issue).
3. Observe how the player changes the video.
4. Repeat steps 1 - 2, but after the first video begins to play put the app into the background by going to the home screen or locking the device.
5. Observe how the player does not automatically play the next video.
6. Open the app up, observe how the player immediately begins playing the next video source.

Describe what you expected to happen:

1. In either foreground or background mode, the video should change when the source prop updates.

## Reproducible sample code

```
const [uri, setUri] = useState("https://myvideo.com/one.mpeg");

setTimeout(() => setUri("https://myvideo.com/two.mpeg"), 10000);

return <Video
    ignoreSilentSwitch="ignore"
    playInBackground
    playWhenInactive
    source={{ uri }}
    onLoad={handleLoad}
    onLoadStart={handleLoadStart}
/>
```

**Comment by rorygarand** [comment]
Does anyone have any ideas? Is this intended behavior?

**Comment by CHaNGeTe** [comment]
Probably your setTimeout is not even firing in background, can you check it?
If that is the case, it is out of scope of the video player.

You can try with approaches like:
https://www.npmjs.com/package/react-native-background-task
https://www.npmjs.com/package/react-native-background-job

**Comment by rorygarand** [comment]
@CHaNGeTe The timeout is definitely firing in the background. I also confirmed through using the available methods, as well as logging within the ios code as well, that the video is loading but strictly not playing.

It's really not a factor of the task being fired or not so those libraries don't really address the problem.

The provided code was just a way to easily replicate the issue but it's not the underlying use case.

I want a queue of videos to play. I've already implemented a way to do this and it works, except the next video won't start playing until the user returns the app to the foreground.

Will you confirm with me that implementing a queue of videos is outside the scope of this project?

**Comment by CHaNGeTe** [comment]
No after the info you provided. I will reopen the issue then. Just wanted confirmation of the js part firing.

**Comment by CHaNGeTe** [comment]
I will check but probably the modifier is not being applied when new source is set.
Can you try this workaround?
add a key prop with uri as value.
`key={uri}`

Not sure if it will do the trick, but can you check in the meanwhile?

**Comment by rorygarand** [comment]
Sure. I added the `key` prop and I am seeing the same behavior.

Also, I'm more than happy to put a PR if I can find a fix to this. I'm far better at javascript so I haven't had much luck pinning down what / where the problem is.

If you are able to provide an idea of what isn't working here I'll start poking around.

**Comment by rorygarand** [comment]
@CHaNGeTe any updates on this? point me in the right direction and i'll put in a PR.

**Comment by rorygarand** [comment]
@CHaNGeTe ?

**Comment by CHaNGeTe** [comment]
Sorry I have been a bit away.
I have no experience in background behaviours on iOS, but I will try to get a look.

It is totally unrelated, but try this PR just to make sure player is released correctly:
https://github.com/react-native-community/react-native-video/pull/1787

I have a though, but dunno if it is the reason, when setting the player we are using main queue:
https://github.com/react-native-community/react-native-video/blob/master/ios/Video/RCTVideo.m#L350

Maybe you can try to fire this piece of code on a background queue to see if it works when in background mode.

**Comment by rorygarand** [comment]
I understand, I very much appreciate you getting back to me.

I'll play around with the PR and check out that area of the code shortly and report back.

Thank you!

**Comment by CHaNGeTe** [comment]
Thanks for your patience @rorygarand I will try to take a look in parallel if I find a slot

**Comment by eddyjusufi** [comment]
@CHaNGeTe @rorygarand Has there been any progress so far? I am also experiencing this.

**Comment by rorygarand** [comment]
Hey @EddieEJ i still have not attempted this. I'll post up here as soon as I've tested.

**Comment by rorygarand** [comment]
@CHaNGeTe 

I checked out https://github.com/react-native-community/react-native-video/pull/1787. While it didn't fix this particular issue, it didn't appear to break anything else.

I tried this inside `- (void)setSrc:(NSDictionary *)source`, but maybe it wasn't exactly what you were suggesting:

```
dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
```

instead of

```
dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t) 0), dispatch_get_main_queue(), ^{
```

I experienced the same behavior.

**Comment by rorygarand** [comment]
@CHaNGeTe have you had a chance to investigate this?

**Comment by polnikale** [comment]
Hey @rorygarand. Have you had a chance to fix this issue?

**Comment by rorygarand** [comment]
I tried a couple of times but my Objective C / Swift chops are pretty non-existent. The video element wasn't really what was important for transitions when backgrounded, so we moved the audio into [react-native-track-player](https://github.com/react-native-kit/react-native-track-player).

I do know part of the problem that is solved in this other library: their queuing management. iOS shuts down the process as soon as the audio stops. If it knows that there's more audio queued up, then it won't.

I haven't checked on this library in a long time but last year it didn't provide any queue management, and I think that was the underlying issue.

**Comment by polnikale** [comment]
@rorygarand thanks for the input!

**Comment by polnikale** [comment]
I've seems like been able to fix the issue. 
The issue was that it worked OK, but `[AVPlayer play]` didn't work at all when the device was locked no matter what. After some digging I was able to find a question on StackOverflow, mentioning how to fix this behavior https://stackoverflow.com/questions/11515883/playing-avaudio-from-ipod-library-while-device-is-locked
We need to manually call
`[[UIApplication sharedApplication] beginReceivingRemoteControlEvents];` to be able to control this `[AVPlayer play]` when device is locked. Also, obviously, we need to call `[[UIApplication sharedApplication] endReceivingRemoteControlEvents]`. I've put these lines into `didEnterBackground/willEnterForeground`.
It's also important to add 
```
- (BOOL)canBecomeFirstResponder {
    return YES;
}
```
I'm not sure if it breaks anything. On our side, everything seems to work as before. 
@CHaNGeTe what do you think about this? Is it going to break anything? If not, I would be pleased to confirm a PR to fix this annoying bug :)

**Comment by tannguyen05** [comment]
> I've seems like been able to fix the issue.
> The issue was that it worked OK, but `[AVPlayer play]` didn't work at all when the device was locked no matter what. After some digging I was able to find a question on StackOverflow, mentioning how to fix this behavior https://stackoverflow.com/questions/11515883/playing-avaudio-from-ipod-library-while-device-is-locked
> We need to manually call
> `[[UIApplication sharedApplication] beginReceivingRemoteControlEvents];` to be able to control this `[AVPlayer play]` when device is locked. Also, obviously, we need to call `[[UIApplication sharedApplication] endReceivingRemoteControlEvents]`. I've put these lines into `didEnterBackground/willEnterForeground`.
> It's also important to add
> 
> ```
> - (BOOL)canBecomeFirstResponder {
>     return YES;
> }
> ```
> 
> I'm not sure if it breaks anything. On our side, everything seems to work as before.
> @CHaNGeTe what do you think about this? Is it going to break anything? If not, I would be pleased to confirm a PR to fix this annoying bug :)

i have same issues. Can you provide some code?pls

**Comment by polnikale** [comment]
Hey @tannguyen05. Open `RCTVideo.m`
Add this code somewhere inside the class(before "end")
```
- (BOOL)canBecomeFirstResponder {
    return YES;
}
```
Also, check out the applicationDidEnterBackground, applicationWillEnterForeground methods and add 1 line:

```
- (void)applicationDidEnterBackground:(NSNotification *)notification
{
  [[UIApplication sharedApplication] beginReceivingRemoteControlEvents]; // add this line!
  if (_playInBackground) {
    // Needed to play sound in background. See https://developer.apple.com/library/ios/qa/qa1668/_index.html
    [_playerLayer setPlayer:nil];
  }
}

- (void)applicationWillEnterForeground:(NSNotification *)notification
{
  [[UIApplication sharedApplication] endReceivingRemoteControlEvents]; // add this line!
  [self applyModifiers];
  if (_playInBackground) {
    [_playerLayer setPlayer:_player];
  }
}
```
It should look something like this
![image](https://user-images.githubusercontent.com/26389743/102075579-1d2f4200-3e0f-11eb-9c50-bb4e4ef28d10.png)
Let me know if this helped :)

**Comment by catmans1** [comment]
@polnikale It work well, it should be merged :)

**Comment by kitsbits** [comment]
@CHaNGeTe @polnikale Any updates on this one? Would love this to be merged, if possible.

**Comment by polnikale** [comment]
I will submit a PR this weekend :)

**Comment by gaurav-cultfit** [comment]
@CHaNGeTe @polnikale Is the change merged ? Because I'm still facing this issue.

**Comment by polnikale** [comment]
@gaurav-cultfit as far as I can see, it's not merged. However, you can install the react-native-dependency from here: https://github.com/polnikale/react-native-video in case you really need it :)

**Comment by gaurav-cultfit** [comment]
@polnikale This still doesn't work. I used the dependency that you provided. Still the issue isn't getting resolved for me. Please help

**Comment by polnikale** [comment]
@gaurav-cultfit could you share what exactly you see as the result? Have you reinstalled Pods after installing the dependency?

**Comment by duansailong** [comment]
[fix ios video not playing in background](https://github.com/polnikale/react-native-video/commit/64d01143a85f2c5cb0a0150831896cb23b9a65d2)

how to fix this question in android   @polnikale
